File: 프로그래머스/1/172928. 공원 산책/공원 산책.java
class Solution {
    static void move(String[] park, int[] start, String direction, int len){
        if(direction.equals("N")){
            for(int i = 1; i <= len; i++){
                if(start[0] - i < 0 || park[start[0] - i].charAt(start[1]) == 'X') {
                    return;
                }
            }
            
            start[0] = start[0] - len;
        }
        else if(direction.equals("E")){
            for(int i = 1; i <= len; i++){
                if(start[1] + i >= park[0].length() || park[start[0]].charAt(start[1] + i) == 'X') {
                    return;
                }
            }
            
            start[1] = start[1] + len;
        }
        else if(direction.equals("S")){
            for(int i = 1; i <= len; i++){
                if(start[0] + i >= park.length || park[start[0] + i].charAt(start[1]) == 'X') {
                    return;
                }
            }
            
            start[0] = start[0] + len;
        }
        else if(direction.equals("W")){
            for(int i = 1; i <= len; i++){
                if(start[1] - i < 0 || park[start[0]].charAt(start[1] - i) == 'X') {
                    return;
                }
            }
            
            start[1] = start[1] - len;
        }
    }
    
    static void findStart(String[] park, int[] start){
        for(int i = 0; i < park.length; i++){
            for(int j = 0; j < park[0].length(); j++){
                if(park[i].charAt(j) == 'S'){
                    start[0] = i;
                    start[1] = j;
                    return;
                }
            }
        }
    }
    
    public int[] solution(String[] park, String[] routes) {
        int[] answer = new int[2];
        int[] start = new int[2];
        findStart(park, start);
        
        for(String route : routes){
            String[] r = route.split(" ");
            String direction = r[0];
            int len = Integer.parseInt(r[1]);
            move(park, start, direction, len);
        }
        
        answer[0] = start[0];
        answer[1] = start[1];
        
        return answer;
    }
}

File: 프로그래머스/1/172928. 공원 산책/공원 산책.py
def solution(park, routes):
    n, m = len(park), len(park[0])
    dx = {'N':-1, 'S':1, 'E':0, 'W': 0}
    dy = {'N': 0, 'S':0, 'E':1, 'W':-1}
    
    for i in range(n):
        for j in range(m):
            if park[i][j] == 'S':
                x = i
                y = j
    
    def check(x, y, dire, dist):
        for i in range(1, dist + 1):
            nx = x + (dx[dire] * i)
            ny = y + (dy[dire] * i)
            
            if not (0 <= nx < n and 0 <= ny < m) or park[nx][ny] == 'X':
                return (x, y)
                
    
        x += (dx[dire] * dist)
        y += (dy[dire] * dist)
        
        return (x, y)
            
        
    for route in routes:
        dire, dist = route.split(" ")
        x, y = check(x, y, dire, int(dist))
            
    
    return [x, y]